### PR TITLE
[WIP][FIX] Bugfix dummy decoder

### LINKED ISSUE #2768: Decoder with dummy classifier estimator always predict same label
It looks like the decoder with dummy classifier is always predicting the first class no matter the frequencies of the labels.

Here is an example using the Haxby dataset, where I change the target frequencies:

```python
import numpy as np
import pandas as pd
from numpy.random import choice
from collections import Counter
from nilearn.image import index_img
from nilearn.datasets import fetch_haxby
from nilearn.decoding import Decoder

# Load and pre-process data
haxby_dataset = fetch_haxby()
behavioral = pd.read_csv(haxby_dataset.session_target[0], delimiter=' ')
conditions = behavioral['labels']
# We consider 4 classes here
condition_mask = conditions.isin(['face', 'cat', 'bottle', 'shoe'])
data_subset = index_img(haxby_dataset.func[0], 
                        condition_mask)
conditions = conditions[condition_mask].values
print("Classes are initially perfectly balanced:")
print(Counter(conditions))
# Split data in train test for simplicity
train = index_img(data_subset, slice(0, -20)) 
test  = index_img(data_subset, slice(-20, None))
conditions_train = conditions[:-20]
conditions_test = conditions[-20:]
print("Dummy decoder predictions:")
dummy_decoder = Decoder(estimator='dummy_classifier', 
                        mask=haxby_dataset.mask_vt[0])
dummy_decoder.fit(train, 
                  conditions_train)
prediction = dummy_decoder.predict(test)
dummy_decoder.classes_[prediction]
```
```
Classes are initially perfectly balanced:
Counter({'face': 108, 'cat': 108, 'shoe': 108, 'bottle': 108})
Dummy decoder predictions:
array(['bottle', 'bottle', 'bottle', 'bottle', 'bottle', 'bottle',
       'bottle', 'bottle', 'bottle', 'bottle', 'bottle', 'bottle',
       'bottle', 'bottle', 'bottle', 'bottle', 'bottle', 'bottle',
       'bottle', 'bottle'], dtype='<U6')
```

```python
# Resample the labels to make bottle very unfrequent
# and face very frequent, such that we would expect the
# dummy decoder to always predict face rather than bottle
fake_conditions = choice(['face', 'cat', 'bottle', 'shoe'], 
                         len(conditions), 
                         p=[0.6, 0.2, 0.05, 0.15])
print("Resample the labels to make bottle very unlikely:")
print(Counter(fake_conditions))
fake_conditions_train = fake_conditions[:-20]
fake_conditions_test = fake_conditions[-20:]
print("Dummy decoder predictions:")
dummy_decoder = Decoder(estimator='dummy_classifier', 
                        mask=haxby_dataset.mask_vt[0])
dummy_decoder.fit(train, 
                  fake_conditions_train)
prediction = dummy_decoder.predict(test)
dummy_decoder.classes_[prediction]
```
```
Resample the labels to make bottle very unlikely:
Counter({'face': 265, 'cat': 93, 'shoe': 55, 'bottle': 19})
Dummy decoder predictions:
array(['bottle', 'bottle', 'bottle', 'bottle', 'bottle', 'bottle',
       'bottle', 'bottle', 'bottle', 'bottle', 'bottle', 'bottle',
       'bottle', 'bottle', 'bottle', 'bottle', 'bottle', 'bottle',
       'bottle', 'bottle'], dtype='<U6')
```

Maybe I'm missing something here, but I would expect a behavior like the following:

```python
from nilearn.input_data import NiftiMasker
from sklearn.dummy import DummyClassifier
d = DummyClassifier()
masker = NiftiMasker(haxby_dataset.mask_vt[0])
X_train = masker.fit_transform(train)
d.fit(X_train, 
      fake_conditions_train)
d.predict(masker.fit_transform(test))
```
```
array(['face', 'face', 'face', 'face', 'face', 'face', 'face', 'face',
       'face', 'face', 'face', 'face', 'face', 'face', 'face', 'face',
       'face', 'face', 'face', 'face'], dtype='<U4')
```

### PR DESCRIPTION
**Pull request by NicolasGensollen**
Fixes #2768 #2767 

WIP...

### CODE REVIEW

**Review by bthirion — COMMENTED** [review]
For this type of PR, you should quickly write a test that fails with previous code by passes with current code.

**Comment by tbng** [comment]
Sorry what's the status of this PR? I saw all the check got passed (barring codecov), but if there is incomplete piece I probably can help.

**Comment by bthirion** [comment]
I think that there is one line to cover by tests
+ make sure that there is a test enforcing the new behavior.

**Comment by NicolasGensollen** [comment]
Hi @tbng 
This PR is still an early draft I started a few days ago but I haven't had the time to work on it since then...
If you want to work on these dummy decoder issues, please do so! I don't think I'll have the time to work on this during the sprint anyway.

**Comment by tbng** [comment]
> Hi @tbng
> This PR is still an early draft I started a few days ago but I haven't had the time to work on it since then...
> If you want to work on these dummy decoder issues, please do so! I don't think I'll have the time to work on this during the sprint anyway.

Thanks, I will try to take over this issue if you don't mind. But then to simplify and not having to open another one should I clone this PR from nilearn upstream, work with my own fork then push it ?

**Comment by NicolasGensollen** [comment]
@tbng I think the easiest way would be for you to open a brand new PR and start from there.
Since this one never went too far, I'll close it.

**Comment by NicolasGensollen** [comment]
Also you can fix #2767 in a self contained PR (not like what I did here...).
The fix should more or less be implemented here.
#2768 is probably a little bit more work.

**Comment by NicolasGensollen** [comment]
@tbng I was expecting to get only one label indeed, but the most frequent one, not systematically the first label of the list. Am I missing something?

**Comment by thomasbazeille** [comment]
I think "stratified" is a fair account in all cases and the recommended one too. Thanks for looking into it Binh!

**Comment by tbng** [comment]
> @tbng I was expecting to get only one label indeed, but the most frequent one, not systematically the first label of the list. Am I missing something?

I think you're right, there seems to be a problem with the `BaseDecoder_predict_dummy`. I am making a fix and ideally would push it with the test case for this in a separate PR.

**Comment by NicolasGensollen** [comment]
Closing this. See continuation in #2826
